HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,46 @@
+from pyspark import SparkContext, AccumulatorParam
+
+def bfs_serial(graph, root, iteration=10):
+    '''
+    Serial implementation of BFS
+    Assuming diameter is 10
+    '''
+    paths = {}
+    current = {root}
+    for step in range(0, iteration+1):
+        next = set()
+        filtered = graph.filter(lambda (k, v): k in current).cache()
+        for n in current:
+            if n not in paths:
+                paths[n] = step
+                next |= filtered.lookup(n)[0]
+        current = next - set(paths.keys())
```

Comment:
This is really elegant, using a set that you can check against.  I couldn't for the life of me how to figure out how to have a collection of the heroes that had already been traversed without using a collect() to get them into a list, so I had some extra steps: .distinct() to get rid of duplicates, subtract() to get rid of the ones that are already in there.  Your version is probably much quicker.